File: face_tracker.py
import socket
import struct
from typing import Callable
from enum import Enum


# enum with values TRACKING and LOST
class State(Enum):
    TRACKING = 0
    LOST = 1


def to_dict(tup: tuple):
    keys = [
        "timestamp",
        "id",
        "width",
        "height",
        "blink_l",
        "blink_r",
        "success",
        "pnp_error",
        "q0",
        "q1",
        "q2",
        "q3",
        "yaw",
        "pitch",
        "roll",
        "x",
        "y",
        "z",
    ]

    return {key: value for key, value in zip(keys, tup)}


class FaceTracker:
    def __init__(self, address: str, port: int) -> None:
        self.server_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.server_socket.settimeout(15)
        self.address = address
        self.port = port

        self.on_data: Callable[[dict], None] | None = None
        self.on_face_lost: Callable[[], None] | None = None
        self.on_timeout: Callable[[], None] | None = None
        self.state = State.LOST

    def start(self):
        self.server_socket.bind((self.address, self.port))

        while True:
            try:
                message, _ = self.server_socket.recvfrom(2048)
            except TimeoutError:
                if self.state == State.TRACKING:
                    self.server_socket.settimeout(15)
                    self.state = State.LOST
                    if self.on_face_lost is not None:
                        self.on_face_lost()
                elif self.state == State.LOST:
                    self.state = State.LOST
                    self.server_socket.settimeout(None)
                    if self.on_timeout is not None:
                        self.on_timeout()
                continue

            self.state = State.TRACKING
            self.server_socket.settimeout(0.2)
            unpacked_data = struct.unpack(
                ("di" + ("f" * 4) + "B"),
                message[:29],
            )
            unpacked_data += struct.unpack("f" * 11, message[29:73])

            data = to_dict(unpacked_data)

            if self.on_data is not None:
                self.on_data(data)

File: src/face_tracker.rs
use std::{io::Error, mem::size_of, time::Duration};

enum State {
    Tracking,
    Idle,
}

#[repr(packed)]
#[derive(Debug, Clone)]
pub struct FaceTrackerData {
    pub timestamp: f64,
    pub face_id: i32,
    pub width: f32,
    pub height: f32,
    pub blink_l: f32,
    pub blink_r: f32,
    pub success: u8,
    pub pnp_error: f32,
    pub q0: f32,
    pub q1: f32,
    pub q2: f32,
    pub q3: f32,
    pub yaw: f32,
    pub pitch: f32,
    pub roll: f32,
    pub y: f32,
    pub x: f32,
    pub z: f32,
}

impl FaceTrackerData {
    fn from_bytes(bytes: &[u8; size_of::<Self>()]) -> Self {
        unsafe { std::mem::transmute(*bytes) }
    }
}

pub struct FaceTracker<DataCallback, FaceLostCallback, TimeoutCallback>
where
    DataCallback: FnMut(&FaceTrackerData),
    FaceLostCallback: FnMut(),
    TimeoutCallback: FnMut(),
{
    socket: std::net::UdpSocket,
    state: State,
    pub on_data: DataCallback,
    pub on_face_lost: FaceLostCallback,
    pub on_timeout: TimeoutCallback,
}

impl<DataCallback, FaceLostCallback, TimeoutCallback>
    FaceTracker<DataCallback, FaceLostCallback, TimeoutCallback>
where
    DataCallback: FnMut(&FaceTrackerData),
    FaceLostCallback: FnMut(),
    TimeoutCallback: FnMut(),
{
    pub fn new(
        address: &str,
        on_data: DataCallback,
        on_face_lost: FaceLostCallback,
        on_timeout: TimeoutCallback,
    ) -> Self {
        let socket = std::net::UdpSocket::bind(address).expect("Unable to connect");
        socket.set_nonblocking(false).unwrap();
        socket
            .set_read_timeout(Some(Duration::from_secs(15)))
            .unwrap();
        Self {
            socket,
            state: State::Idle,
            on_data,
            on_face_lost,
            on_timeout,
        }
    }

    pub fn start(&mut self) {
        let mut buf = [0; std::mem::size_of::<FaceTrackerData>()];
        loop {
            match self.socket.recv_from(&mut buf) {
                Ok(_) => {
                    self.handle_data_recv(&buf);
                }
                Err(err) => {
                    self.handle_err(&err);
                }
            }
        }
    }

    fn handle_data_recv(&mut self, buf: &[u8; size_of::<FaceTrackerData>()]) {
        self.state = State::Tracking;
        self.socket
            .set_read_timeout(Some(Duration::from_millis(200)))
            .unwrap();
        let data = FaceTrackerData::from_bytes(buf);
        (self.on_data)(&data);
    }

    fn handle_err(&mut self, error: &Error) {
        if error.kind() != std::io::ErrorKind::TimedOut
            && error.kind() != std::io::ErrorKind::WouldBlock
        {
            panic!("{error:?}");
        }
        match self.state {
            State::Tracking => {
                self.state = State::Idle;
                self.socket
                    .set_read_timeout(Some(Duration::from_secs(15)))
                    .unwrap();
                (self.on_face_lost)()
            }
            State::Idle => {
                self.socket.set_read_timeout(None).unwrap();
                (self.on_timeout)()
            }
        }
    }
}
